File: Sources/ManageArmsCore/Fetcher.swift
import Foundation
import Synchronization
import Darwin

/// 取得した中身の解釈結果。**自動では入れない。**
/// README からのコマンド抽出は必ず外すため、確認画面を挟む（DESIGN.md 6 章）。
public struct Candidate: Equatable, Sendable {
    public let kind: Kind
    public let name: String
    public let description: String?
    /// 一時ディレクトリ内の実体。`Staging.discard()` か install で片付く。
    public let localURL: URL
}

/// 一時展開の所有権。install するか discard するまで生きている。
public final class Staging: @unchecked Sendable {
    public let root: URL
    public let source: GitHubSource
    public let candidates: [Candidate]
    public let resolvedSHA: String?

    private let lock = NSLock()
    private var applying = false
    private var discarded = false

    public init(root: URL, source: GitHubSource, candidates: [Candidate],
                resolvedSHA: String? = nil) {
        self.root = root
        self.source = source
        self.candidates = candidates
        self.resolvedSHA = resolvedSHA
    }

    func claimForApply() -> Bool {
        lock.withLock {
            guard !applying, !discarded else { return false }
            applying = true
            return true
        }
    }

    func releaseAfterFailure() { lock.withLock { applying = false } }

    func finishApply() {
        let shouldRemove = lock.withLock {
            applying = false
            guard !discarded else { return false }
            discarded = true
            return true
        }
        if shouldRemove { try? FileManager.default.removeItem(at: root) }
    }

    /// 一時ディレクトリは `temporaryDirectory` 配下なので OS も回収するが、
    /// アプリが自前の掃除機能を持たなくて済むよう明示的に消す（9 章）。
    public func discard() {
        let shouldRemove = lock.withLock {
            guard !applying, !discarded else { return false }
            discarded = true
            return true
        }
        if shouldRemove { try? FileManager.default.removeItem(at: root) }
    }
}

public enum Fetcher {

    /// monorepo の zipball は subdir が 20 KB でも数百 MB になり得る（3.3）。
    public static let sizeLimit = 50 * 1024 * 1024

    public enum Failure: Error, Equatable, CustomStringConvertible {
        case tooLarge(repo: String, bytes: Int)
        case downloadFailed(String)
        case extractFailed(String)
        case subdirNotFound(String)
        case nothingRecognized

        public var description: String {
            switch self {
            case .tooLarge(let repo, let bytes):
                String(localized: "\(repo) のアーカイブが大きすぎます（\(bytes / 1024 / 1024) MB、上限 50 MB）")
            case .downloadFailed(let m):
                String(localized: "ダウンロードに失敗しました: \(m)")
            case .extractFailed(let m):
                String(localized: "展開に失敗しました: \(m)")
            case .subdirNotFound(let p):
                String(localized: "\(p) がアーカイブ内に見つかりません")
            case .nothingRecognized:
                String(localized: "SKILL.md も plugin.json も見つかりません。対応していない形式です")
            }
        }
    }

    /// zip を落として展開し、中身から種別を判定する。`git clone` は使わない（3.3）。
    public static func stage(
        _ source: GitHubSource,
        defaultBranch: String = "main",
        resolvedSHA: String? = nil
    ) async throws -> Staging {
        let root = URL(filePath: NSTemporaryDirectory())
            .appending(path: "manage-arms-fetch-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)

        do {
            let zip = root.appending(path: "archive.zip")
            try await download(source.archiveURL(defaultBranch: defaultBranch,
                                                 revision: resolvedSHA),
                               to: zip, repo: source.repo)
            let available = try? root.resourceValues(
                forKeys: [.volumeAvailableCapacityForImportantUsageKey])
                .volumeAvailableCapacityForImportantUsage
            guard available == nil || available! >= Int64(extractedSizeLimit * 2) else {
                throw Failure.extractFailed(String(localized: "アーカイブを安全に展開する空き容量がありません"))
            }

            let unpacked = root.appending(path: "unpacked")
            try await extract(zip, to: unpacked)
            try? FileManager.default.removeItem(at: zip)   // zip 本体はもう要らない

            // zipball は `<repo>-<branch>/` を 1 段かぶせる
            let top = try singleTopLevel(of: unpacked)
            let base = source.subdir.map { top.appending(path: $0) } ?? top
            guard FileManager.default.fileExists(atPath: base.path(percentEncoded: false)) else {
                throw Failure.subdirNotFound(source.subdir ?? "/")
            }

            // ディレクトリ名由来の候補も含めて、最後にもう一度名前を検査する。
            // ここを通った名前だけが `Installer` でパスに使われる（9 章の作成方向）。
            let candidates = identify(base).filter { WriteGuard.isValidName($0.name) }
            guard !candidates.isEmpty else { throw Failure.nothingRecognized }
            return Staging(root: root, source: source, candidates: candidates,
                           resolvedSHA: resolvedSHA)
        } catch {
            try? FileManager.default.removeItem(at: root)
            throw error
        }
    }

    static let extractedSizeLimit = 200 * 1024 * 1024
    static let singleFileLimit = 20 * 1024 * 1024
    static let entryLimit = 10_000
    static let extractTimeLimit: TimeInterval = 30

    /// `strict` はこれから利用者のディレクトリへ入る木にだけ使う。アーカイブ全体の
    /// 走査では上限だけを見る — 取り出さない場所（リポジトリ直下の `AGENTS.md` が
    /// symlink 等）まで拒むと、subdir を指した取得まで巻き添えで失敗する。
    static func validateExtractedTree(_ root: URL, strict: Bool = true) throws {
        let keys: [URLResourceKey] = [.isRegularFileKey, .isDirectoryKey,
                                      .isSymbolicLinkKey, .fileSizeKey]
        guard let walker = FileManager.default.enumerator(
            at: root, includingPropertiesForKeys: keys,
            options: [], errorHandler: { _, _ in false }) else { return }
        var entries = 0
        var total = 0
        for case let url as URL in walker {
            entries += 1
            guard entries <= entryLimit else {
                throw Failure.extractFailed(String(localized: "展開したファイル数が上限を超えました"))
            }
            let values = try url.resourceValues(forKeys: Set(keys))
            let unsupported = values.isSymbolicLink == true
                || !(values.isDirectory == true || values.isRegularFile == true)
            if unsupported {
                guard strict else { continue }   // 列挙は symlink を降りない
                throw Failure.extractFailed(values.isSymbolicLink == true
                    ? String(localized: "アーカイブにシンボリックリンクが含まれています")
                    : String(localized: "アーカイブに対応していない種類のファイルが含まれています"))
            }
            if values.isRegularFile == true {
                let size = values.fileSize ?? 0
                guard size <= singleFileLimit else {
                    throw Failure.extractFailed(String(localized: "展開したファイルが大きすぎます"))
                }
                total += size
                guard total <= extractedSizeLimit else {
                    throw Failure.extractFailed(String(localized: "展開後の合計サイズが上限を超えました"))
                }
            }
        }
    }

    static func validate(_ candidate: Candidate, in staging: Staging) throws {
        let stagedRoot = staging.root.resolvingSymlinksInPath().standardizedFileURL
        let local = candidate.localURL.resolvingSymlinksInPath().standardizedFileURL
        guard local.path == stagedRoot.path || WriteGuard.isInside(local, stagedRoot),
              !WriteGuard.isSymlink(candidate.localURL) else {
            throw Failure.extractFailed(String(localized: "取得物が一時領域の外を参照しています"))
        }
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: candidate.localURL.path,
                                             isDirectory: &isDirectory) else {
            throw Failure.extractFailed(String(localized: "取得物が見つかりません"))
        }
        if isDirectory.boolValue {
            try validateExtractedTree(candidate.localURL)
        } else {
            let values = try candidate.localURL.resourceValues(
                forKeys: [.isRegularFileKey, .fileSizeKey])
            guard values.isRegularFile == true,
                  (values.fileSize ?? 0) <= singleFileLimit else {
                throw Failure.extractFailed(String(localized: "取得物のファイルが大きすぎるか、対応していない形式です"))
            }
        }
    }

    // MARK: - ダウンロード

    /// codeload は GET には `Content-Length` を返す（HEAD には返さないので
    /// `curl -I` では見えない）。実測では最初の進捗コールバックで全体サイズが判明し、
    /// **10 KB 書いた時点で中断**できる。返らない場合の保険として
    /// 書き込み量による中断も残してある（3.3）。
    ///
    /// ⚠️ completion handler 付きの `downloadTask` は
    /// デリゲートの進捗コールバックを無効化する。中断するには
    /// デリゲート駆動にして継続を自分で resume する必要がある。
    static func download(_ url: URL, to destination: URL, repo: String) async throws {
        let downloader = Downloader(limit: sizeLimit, destination: destination, repo: repo)
        // ephemeral: ディスクキャッシュを持たない（9 章「キャッシュディレクトリを持たない」）。
        let session = URLSession(configuration: .ephemeral, delegate: downloader,
                                 delegateQueue: nil)
        defer { session.finishTasksAndInvalidate() }
        try await downloader.run(session: session, url: url)
    }

    final class Downloader: NSObject, URLSessionDownloadDelegate, @unchecked Sendable {
        let limit: Int
        let destination: URL
        let repo: String

        private let lock = NSLock()
        private var continuation: CheckedContinuation<Void, any Error>?
        private var exceeded = false
        private var written = 0
        /// 判明していればアーカイブ全体のサイズ。エラー表示に使う。
        private var expectedTotal = 0
        private var stagingError: (any Error)?

        init(limit: Int, destination: URL, repo: String) {
            self.limit = limit; self.destination = destination; self.repo = repo
        }

        func run(session: URLSession, url: URL) async throws {
            let task = session.downloadTask(with: url)
            try await withTaskCancellationHandler {
                try await withCheckedThrowingContinuation { cont in
                    lock.withLock { continuation = cont }
                    if Task.isCancelled { task.cancel() } else { task.resume() }
                }
            } onCancel: {
                task.cancel()
            }
        }

        private func finish(_ result: Result<Void, any Error>) {
            let cont = lock.withLock { () -> CheckedContinuation<Void, any Error>? in
                defer { continuation = nil }
                return continuation
            }
            cont?.resume(with: result)
        }

        func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                        didWriteData bytesWritten: Int64, totalBytesWritten: Int64,
                        totalBytesExpectedToWrite expected: Int64) {
            let over = (expected > 0 && expected > Int64(limit)) || totalBytesWritten > Int64(limit)
            lock.withLock {
                written = Int(totalBytesWritten)
                if expected > 0 { expectedTotal = Int(expected) }
                if over { exceeded = true }
            }
            if over { downloadTask.cancel() }
        }

        /// 受け取った一時ファイルはこの関数を抜けると消えるので、ここで移す。
        func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                        didFinishDownloadingTo location: URL) {
            do {
                if let http = downloadTask.response as? HTTPURLResponse, http.statusCode != 200 {
                    throw Failure.downloadFailed("HTTP \(http.statusCode)")
                }
                try FileManager.default.moveItem(at: location, to: destination)
            } catch {
                lock.withLock { stagingError = error }
            }
        }

        func urlSession(_ session: URLSession, task: URLSessionTask,
                        didCompleteWithError error: (any Error)?) {
            let (over, bytes, staged) = lock.withLock {
                (exceeded, max(expectedTotal, written), stagingError)
            }
            if over {
                finish(.failure(Failure.tooLarge(repo: repo, bytes: bytes)))
            } else if let staged {
                finish(.failure(staged))
            } else if let error {
                finish(.failure(Failure.downloadFailed(error.localizedDescription)))
            } else {
                finish(.success(()))
            }
        }
    }

    // MARK: - 展開

    /// Foundation に zip 展開 API は無い。ZIPFoundation 等の依存を足すより
    /// OS 同梱の `ditto` に任せる方がメモリにも載らず依存もゼロ（3.3）。
    /// Zip Slip 安全であることは spike #13 で実測済み。
    /// **`async` なのは待ち方のため。** 以前は `Thread.sleep` で回していたが、
    /// ここは nonisolated async の中（`stage` から呼ばれる）なので、
    /// グローバル executor のスレッドを最大 30 秒占有することになる。
    /// 協調スレッドプールはコア数程度しか無いので、塞ぐのではなく返す。
    static func extract(_ zip: URL, to destination: URL) async throws {
        try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: true)
        let process = Process()
        process.executableURL = URL(filePath: "/usr/bin/ditto")
        process.arguments = ["-xk", zip.path(percentEncoded: false),
                             destination.path(percentEncoded: false)]
        // **読まないパイプを渡さない。** バッファが埋まると子プロセスが書き込みで止まる。
        // かといって捨てると展開失敗の理由が出せなくなるので、`Exec.run` と同じく
        // 読み手を付けて上限まで溜める（`ditto -xk` は標準出力を使わない）。
        let err = Pipe()
        let message = Mutex(Data())
        // `Exec.run` と同じ理由で EOF を待てるようにする。終了を待っただけでは
        // 未配送分が残り、ハンドラを外した瞬間に消える（失敗理由が空になる）。
        let drained = DispatchGroup()
        drained.enter()
        err.fileHandleForReading.readabilityHandler = { handle in
            let chunk = handle.availableData
            guard !chunk.isEmpty else {
                handle.readabilityHandler = nil
                drained.leave()
                return
            }
            message.withLock { if $0.count < 4096 { $0.append(chunk.prefix(4096 - $0.count)) } }
        }
        defer {
            err.fileHandleForReading.readabilityHandler = nil
            try? err.fileHandleForReading.close()
        }
        process.standardError = err
        process.standardOutput = FileHandle.nullDevice
        try process.run()
        let deadline = Date().addingTimeInterval(extractTimeLimit)
        var validationFailure: (any Error)?
        while process.isRunning {
            if Date() >= deadline {
                validationFailure = Failure.extractFailed(
                    String(localized: "アーカイブの展開が制限時間を超えました"))
                process.terminate()
                break
            }
            do { try validateExtractedTree(destination, strict: false) }
            catch { validationFailure = error; process.terminate(); break }
            // 1 秒間隔。この検査は展開ツリーを毎回歩くので、細かく回すと
            // 上限いっぱい（10,000 項目 × 30 秒）で数十万回の stat になる。
            // ディスクを埋めさせない役目は取得前の空き容量検査が担っている。
            do { try await Task.sleep(for: .seconds(1)) }
            catch {
                // **キャンセルは握り潰さない。** `try?` にすると待たずに回り続け、
                // 上限まで CPU を焼く。畳んで `ditto` を孤児にしない。
                validationFailure = error
                process.terminate()
                break
            }
        }
        if validationFailure != nil, process.isRunning {
            try? await Task.sleep(for: .milliseconds(200))
            if process.isRunning { Darwin.kill(process.processIdentifier, SIGKILL) }
        }
        process.waitUntilExit()
        _ = drained.wait(timeout: .now() + Exec.drainTimeout)
        if let validationFailure { throw validationFailure }
        guard process.terminationStatus == 0 else {
            let reason = String(decoding: message.withLock { $0 }, as: UTF8.self)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            throw Failure.extractFailed(reason.isEmpty
                ? String(localized: "展開コマンドが失敗しました") : reason)
        }
        try validateExtractedTree(destination, strict: false)
    }

    /// zipball は `<repo>-<branch>/` を 1 段かぶせる。それを剥がす。
    static func singleTopLevel(of dir: URL) throws -> URL {
        let entries = (try? FileManager.default.contentsOfDirectory(
            atPath: dir.path(percentEncoded: false)))?.filter { !$0.hasPrefix(".") } ?? []
        guard entries.count == 1 else { return dir }
        return dir.appending(path: entries[0])
    }

    // MARK: - 種別判定

    /// 取得した中身を見て決める。README のテキストからは推測しない（6 章）。
    ///
    /// **symlink は追わない。** `stage()` はアーカイブ全体を strict 検査しなくなった
    /// （取り出さない場所の symlink で導入ごと落とさないため）ので、frontmatter を
    /// 読む前にここで弾く。追うと取得先の細工 1 つで任意のローカルファイルの中身が
    /// 確認画面に出る。実体として入るものは `validate` が改めて strict 検査する。
    static func identify(_ base: URL) -> [Candidate] {
        guard !WriteGuard.isSymlink(base) else { return [] }
        let fm = FileManager.default
        func exists(_ path: String) -> Bool {
            let url = base.appending(path: path)
            return fm.fileExists(atPath: url.path(percentEncoded: false))
                && !WriteGuard.isSymlink(url)
        }

        // marketplace を兼ねたリポジトリは plugin.json と skills/ の両方を持つ。
        // どちらかで打ち切ると片方が選べなくなるので、併記して確認画面で選ばせる（6 章）。
        var found: [Candidate] = []
        if exists(".claude-plugin/plugin.json") {
            found.append(Candidate(kind: .plugin, name: base.lastPathComponent,
                                   description: nil, localURL: base))
        }
        if exists("SKILL.md") {
            let front = FrontmatterParser.read(base.appending(path: "SKILL.md"))
            guard case .parsed(let matter) = front else {
                return found + [Candidate(kind: .skill, name: base.lastPathComponent,
                                          description: nil, localURL: base)]
            }
            // **frontmatter の `name` は取得先が書いた文字列**で、こちらの管理下にない。
            // パス要素として使えない名前（`../` を含む等）はディレクトリ名に落とす。
            let declared = matter.name.flatMap { WriteGuard.isValidName($0) ? $0 : nil }
            return found + [Candidate(kind: .skill, name: declared ?? base.lastPathComponent,
                                      description: matter.description, localURL: base)]
        }

        // `.md` に frontmatter の `tools:` があれば Subagent（6 章）。
        let markdown = (try? fm.contentsOfDirectory(atPath: base.path(percentEncoded: false)))?
            .filter { $0.hasSuffix(".md") && !$0.hasPrefix(".") }.sorted() ?? []
        let subagents = markdown.compactMap { file -> Candidate? in
            let url = base.appending(path: file)
            guard !WriteGuard.isSymlink(url),
                  case .parsed(let matter) = FrontmatterParser.read(url), matter.tools != nil
            else { return nil }
            return Candidate(kind: .subagent, name: String(file.dropLast(3)),
                             description: matter.description, localURL: url)
        }
        if !subagents.isEmpty { return found + subagents }

        return found + skills(under: base, depth: 3)
    }

    /// リポジトリ直下を指された場合、スキルは何段か下に並んでいることがある。
    /// `skills/<name>`（vercel-labs/skills）だけでなく
    /// `skills/<category>/<name>`（mattpocock/skills）もあるので数段たどる。
    /// Subagent の判定は指されたディレクトリ直下だけで行う — 下層の `.md` まで
    /// frontmatter を読むと、ただの文書が候補に混ざる。
    static func skills(under base: URL, depth: Int) -> [Candidate] {
        guard depth > 0 else { return [] }
        let fm = FileManager.default
        let children = (try? fm.contentsOfDirectory(atPath: base.path(percentEncoded: false)))?
            .filter { !$0.hasPrefix(".") }.sorted() ?? []
        return children.flatMap { child -> [Candidate] in
            var isDir: ObjCBool = false
            let url = base.appending(path: child)
            // `fileExists(isDirectory:)` は symlink を追うので、単独では
            // リンクされた実ユーザーのディレクトリまで降りてしまう。
            guard fm.fileExists(atPath: url.path(percentEncoded: false), isDirectory: &isDir),
                  isDir.boolValue, !WriteGuard.isSymlink(url)
            else { return [] }
            if fm.fileExists(atPath: url.appending(path: "SKILL.md")
                                        .path(percentEncoded: false)) {
                return identify(url)
            }
            return skills(under: url, depth: depth - 1)
        }
    }
}

File: ide/fetcher.ts
import {
  createWriteStream, mkdirSync, mkdtempSync, readdirSync, readFileSync, rmSync, statSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, join, resolve, sep } from "node:path";
import { pipeline } from "node:stream/promises";
import { open as openZip, Entry, ZipFile } from "yauzl";
import { KindId } from "../core/agent";
import { AgentToolError } from "../core/errors";
import { safeSegments } from "../core/archive";
import { ENTRY_LIMIT, EXTRACTED_SIZE_LIMIT, PAGE_LIMIT, SINGLE_FILE_LIMIT, SIZE_LIMIT } from "../core/limits";
import * as frontmatter from "./frontmatter";
import { archiveUrl, GitHubSource } from "../core/github";
import { isValidName } from "./writeGuard";

/**
 * 取得した中身の解釈結果。自動では入れない —
 * README からのコマンド抽出は必ず外すので、確認画面を挟む。
 */
export type Candidate = {
  readonly kind: KindId;
  readonly name: string;
  readonly description?: string;
  /** 一時ディレクトリ内の実体。`discard()` か install で片付く。 */
  readonly localPath: string;
  /** Plugin CLI に渡す `plugin@marketplace`。展開ディレクトリ名は使わない。 */
  readonly installSelector?: string;
};

export type Staging = {
  readonly root: string;
  readonly source: GitHubSource;
  readonly candidates: Candidate[];
  readonly resolvedSha?: string;
};

const fail = (message: string): never => {
  throw new AgentToolError("FETCH_FAILED", message);
};

/** 一時領域だけを使う。成功・失敗・キャンセルの全経路で消す。 */
export const discard = (staging: { root: string }): void =>
  rmSync(staging.root, { recursive: true, force: true });

/**
 * カタログページを 1 枚読む。呼び出し側は本文を JSON-LD の抽出にだけ使い、
 * 読み終えたら捨てる（ファイルにも registry にも残さない）。
 */
export async function fetchPage(url: string, fetchImpl: typeof fetch = fetch): Promise<string> {
  const response = await fetchImpl(url, { redirect: "follow", cache: "no-store" });
  if (!response.ok) fail(`the page could not be read: HTTP ${response.status}`);
  const declared = Number(response.headers.get("content-length") ?? "0");
  if (declared > PAGE_LIMIT) {
    await response.body?.cancel();
    fail("the page is too large to read (limit 2 MB)");
  }
  if (!response.body) fail("the page could not be read: empty response");
  const body = response.body!;

  let size = 0;
  let text = "";
  const decoder = new TextDecoder();
  for await (const chunk of body as unknown as AsyncIterable<Uint8Array>) {
    size += chunk.byteLength;
    if (size > PAGE_LIMIT) {
      await body.cancel();
      fail("the page is too large to read (limit 2 MB)");
    }
    text += decoder.decode(chunk, { stream: true });
  }
  return text + decoder.decode();
}

/** zip を落として展開し、中身から種別を判定する。`git clone` は使わない。 */
export async function stage(source: GitHubSource, options: {
  defaultBranch?: string;
  resolvedSha?: string;
  fetchImpl?: typeof fetch;
} = {}): Promise<Staging> {
  const root = mkdtempSync(join(tmpdir(), "agent-tool-fetch-"));
  try {
    const archive = join(root, "archive.zip");
    await download(archiveUrl(source, options.defaultBranch ?? "main", options.resolvedSha),
      archive, source.repo, options.fetchImpl ?? fetch);

    const unpacked = join(root, "unpacked");
    await extract(archive, unpacked);
    rmSync(archive, { force: true });          // zip 本体はもう要らない

    // zipball は `<repo>-<branch>/` を 1 段かぶせる。それを剥がす。
    const top = singleTopLevel(unpacked);
    const base = source.subdir === undefined ? top : join(top, source.subdir);
    if (!isPath(base)) fail(`${source.subdir ?? "/"} was not found in the archive`);

    // ディレクトリ名由来の候補も含めて、最後にもう一度名前を検査する。
    // ここを通った名前だけが `installer` でパスに使われる。
    const candidates = identify(base).filter(candidate => isValidName(candidate.name));
    if (candidates.length === 0) {
      fail("Neither SKILL.md nor plugin.json was found; the format is not supported");
    }
    return { root, source, candidates, resolvedSha: options.resolvedSha };
  } catch (error) {
    rmSync(root, { recursive: true, force: true });
    throw error;
  }
}

/**
 * codeload は GET に `Content-Length` を返す。判明した時点で上限を超えていれば
 * 本文を読まずに切る。返らない場合の保険として、書き込み量でも中断する。
 */
async function download(url: string, destination: string, repo: string,
                        fetchImpl: typeof fetch): Promise<void> {
  const response = await fetchImpl(url, { redirect: "follow", cache: "no-store" });
  if (!response.ok) fail(`download failed: HTTP ${response.status}`);
  const declared = Number(response.headers.get("content-length") ?? "0");
  if (declared > SIZE_LIMIT) {
    await response.body?.cancel();
    fail(`the archive for ${repo} is too large (${Math.round(declared / 1024 / 1024)} MB, limit 50 MB)`);
  }
  if (!response.body) fail("download failed: empty response");

  let written = 0;
  const guard = async function* (source: AsyncIterable<Uint8Array>): AsyncGenerator<Uint8Array> {
    for await (const chunk of source) {
      written += chunk.byteLength;
      if (written > SIZE_LIMIT) {
        fail(`the archive for ${repo} is too large (over 50 MB)`);
      }
      yield chunk;
    }
  };
  await pipeline(response.body as unknown as AsyncIterable<Uint8Array>, guard,
    createWriteStream(destination));
}

/**
 * zip を展開する。**各エントリを書く前に検査する。**
 * ディレクトリ横断（Zip Slip）、symlink、対応しない種別、サイズ上限を
 * ここで落とすので、展開後の木をもう一度歩き直さない。
 */
export function extract(archive: string, destination: string): Promise<void> {
  mkdirSync(destination, { recursive: true });
  return new Promise((done, reject) => {
    openZip(archive, { lazyEntries: true, autoClose: true }, (error, zip: ZipFile) => {
      if (error) return reject(new AgentToolError("FETCH_FAILED", `extraction failed: ${error.message}`));
      let entries = 0;
      let total = 0;
      const abort = (message: string): void => {
        zip.close();
        reject(new AgentToolError("FETCH_FAILED", `extraction failed: ${message}`));
      };
      zip.on("error", (streamError: Error) => abort(streamError.message));
      zip.on("end", () => done());
      zip.on("entry", (entry: Entry) => {
        entries += 1;
        if (entries > ENTRY_LIMIT) return abort("too many files in the archive");

        const target = safeJoin(destination, entry.fileName);
        if (target === null) return abort("the archive escapes the extraction directory");

        // 上位 16 bit が Unix のモード。symlink と特殊ファイルは取り出さない。
        const mode = (entry.externalFileAttributes >>> 16) & 0o170000;
        if (mode === 0o120000) return abort("the archive contains a symbolic link");
        if (mode !== 0 && mode !== 0o100000 && mode !== 0o040000) {
          return abort("the archive contains an unsupported file type");
        }
        if (entry.fileName.endsWith("/")) {
          mkdirSync(target, { recursive: true });
          return zip.readEntry();
        }
        if (entry.uncompressedSize > SINGLE_FILE_LIMIT) return abort("an extracted file is too large");
        total += entry.uncompressedSize;
        if (total > EXTRACTED_SIZE_LIMIT) return abort("the extracted archive is too large");

        zip.openReadStream(entry, (streamError, stream) => {
          if (streamError || !stream) return abort(streamError?.message ?? "entry could not be read");
          mkdirSync(dirname(target), { recursive: true });
          pipeline(stream, createWriteStream(target))
            .then(() => zip.readEntry())
            .catch((writeError: Error) => abort(writeError.message));
        });
      });
      zip.readEntry();
    });
  });
}

/** アーカイブ内のパスを展開先へ落とす。外へ出るものは null。検証は core と共有する。 */
export function safeJoin(root: string, entryName: string): string | null {
  const parts = safeSegments(entryName);
  if (parts === null) return null;
  const target = resolve(root, ...parts);
  return target === resolve(root) || target.startsWith(resolve(root) + sep) ? target : null;
}

const isPath = (path: string): boolean => {
  try {
    statSync(path);
    return true;
  } catch {
    return false;
  }
};

const isDirectory = (path: string): boolean => {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
};

/** zipball は `<repo>-<branch>/` を 1 段かぶせる。 */
export function singleTopLevel(dir: string): string {
  let entries: string[];
  try {
    entries = readdirSync(dir).filter(name => !name.startsWith("."));
  } catch {
    return dir;
  }
  return entries.length === 1 ? join(dir, entries[0]) : dir;
}

/**
 * 取得した中身を見て決める。README のテキストからは推測しない。
 * 展開時に symlink を弾いているので、ここでは種別だけを見る。
 */
export function identify(base: string): Candidate[] {
  const found: Candidate[] = [];

  // marketplace を兼ねたリポジトリは plugin.json と skills/ の両方を持つ。
  // どちらかで打ち切ると片方が選べなくなるので、併記して確認画面で選ばせる。
  if (isPath(join(base, ".claude-plugin", "plugin.json"))) {
    found.push({
      kind: "plugin", name: basename(base), localPath: base,
      installSelector: pluginSelector(base),
    });
  }

  if (isPath(join(base, "SKILL.md"))) {
    const result = frontmatter.read(join(base, "SKILL.md"));
    if (result.status !== "parsed") {
      return [...found, { kind: "skill", name: basename(base), localPath: base }];
    }
    // frontmatter の `name` は取得先が書いた文字列で、こちらの管理下にない。
    // パス要素として使えない名前はディレクトリ名に落とす。
    const declared = result.matter.name !== undefined && isValidName(result.matter.name)
      ? result.matter.name : undefined;
    return [...found, {
      kind: "skill", name: declared ?? basename(base), localPath: base,
      description: result.matter.description,
    }];
  }

  // `.md` に frontmatter の `tools:` があれば Subagent。
  const markdown = (() => {
    try {
      return readdirSync(base).filter(name => name.endsWith(".md") && !name.startsWith(".")).sort();
    } catch {
      return [];
    }
  })();
  const subagents = markdown.flatMap((file): Candidate[] => {
    const result = frontmatter.read(join(base, file));
    if (result.status !== "parsed" || result.matter.tools === undefined) return [];
    return [{
      kind: "subagent", name: file.slice(0, -3), localPath: join(base, file),
      description: result.matter.description,
    }];
  });
  if (subagents.length > 0) return [...found, ...subagents];

  return [...found, ...skillsUnder(base, 3)];
}

function pluginSelector(base: string): string | undefined {
  try {
    const object = JSON.parse(
      readFileSync(join(base, ".claude-plugin", "marketplace.json"), "utf8"),
    ) as { name?: unknown; plugins?: unknown };
    const marketplace = typeof object.name === "string" ? object.name : "";
    const first = Array.isArray(object.plugins) ? object.plugins[0] : undefined;
    const name = first !== null && typeof first === "object" && typeof (first as { name?: unknown }).name === "string"
      ? (first as { name: string }).name : "";
    return name !== "" && marketplace !== "" ? `${name}@${marketplace}` : undefined;
  } catch {
    return undefined;
  }
}

/**
 * リポジトリ直下を指された場合、スキルは何段か下に並んでいることがある。
 * `skills/<name>` だけでなく `skills/<category>/<name>` もあるので数段たどる。
 * Subagent の判定は指されたディレクトリ直下だけで行う — 下層の `.md` まで
 * frontmatter を読むと、ただの文書が候補に混ざる。
 */
export function skillsUnder(base: string, depth: number): Candidate[] {
  if (depth <= 0) return [];
  let children: string[];
  try {
    children = readdirSync(base).filter(name => !name.startsWith(".")).sort();
  } catch {
    return [];
  }
  return children.flatMap(child => {
    const path = join(base, child);
    if (!isDirectory(path)) return [];
    return isPath(join(path, "SKILL.md")) ? identify(path) : skillsUnder(path, depth - 1);
  });
}
